partridge/hmm.py

```diff
@@ -0,0 +1,256 @@
+
+    init_probs = ddict2dict(init_probs)
+    emiss_probs = ddict2dict(emiss_probs)
+    trans_probs = ddict2dict(trans_probs)
+
+    return init_probs, emiss_probs, trans_probs
+
+
+def sparse_viterbi(N, init_probs, emiss_probs, trans_probs):
+    '''
+    Based on the implementation found here: https://git.io/fNQ6S
+    ------------------------------------------------------------
+
+    p = log probability
+    i = stop index
+    j = seg index
+
+    init_probs[i][j] = p
+    trans_probs[i1][j1][i0][j0] = p
+    emiss_probs[i][j] = p
+    '''
+    P = [{} for _ in range(N)]
+    preds = {}
+
+    for seg_idx, eprob in emiss_probs[0].items():
+        if seg_idx in init_probs[0]:
+            iprob = init_probs[0][seg_idx]
+            P[0][seg_idx] = iprob + eprob
+
+    for bstop_idx in range(1, N):
+        if bstop_idx not in emiss_probs:
+            raise NoPathError('HMM break at observation {}'.format(bstop_idx))
+
+        for bseg_idx, eprob in emiss_probs[bstop_idx].items():
+            if (
+                bstop_idx not in trans_probs or
+                bseg_idx not in trans_probs[bstop_idx]
+            ):
+                continue
+
+            for astop_idx, tprobs in trans_probs[bstop_idx][bseg_idx].items():
+                for aseg_idx, tprob in tprobs.items():
+                    if aseg_idx not in P[astop_idx]:
+                        continue
+
+                    curr_prob = P[astop_idx][aseg_idx]
+                    prob = curr_prob + eprob + tprob
+
+                    if (
+                        bseg_idx not in P[bstop_idx] or
+                        prob > P[bstop_idx][bseg_idx]
+                    ):
+                        P[bstop_idx][bseg_idx] = prob
+                        preds[bstop_idx, bseg_idx] = (astop_idx, aseg_idx)
+
+    last_stop_idx = N - 1
+
+    # Find the most probable end state
+    last_seg_idx = dict_argmax(P[-1])
+
+    path = [(last_stop_idx, last_seg_idx)]
+
+    # Reconstruct the most probable path the end state
+    for i in range(last_stop_idx):
+        if path[-1] not in preds:
+            raise NoPathError('HMM break at observation {}'.format(N - i))
+        path.append(preds[path[-1]])
+
+    path.reverse()
+
+    return path
+
+
+class CandidateCache(object):
+    def __init__(self, shape_lons, shape_lats):
+        assert shape_lons.shape[0] > 1
+        assert shape_lons.shape == shape_lats.shape
+        self.shape_lons = shape_lons
+        self.shape_lats = shape_lats
+
+    @cached_property
+    def spatial_index(self):
+        return Index((i, s.bounds, None) for i, s in enumerate(self.segments))
+
+    @cached_property
+    def cumulative_dists(self):
+        return np.cumsum(self.segment_dists)
+
+    @cached_property
+    def segment_dists(self):
+        return pairwise_great_circle_vec(self.shape_lons, self.shape_lats)
+
+    @cached_property
+    def segments(self):
+        return list(map(LineString, pairwise(zip(
```

Comment:
Without digging very deeply, this looks like its creating a bunch of LineStrings that are composed of 2 points (so a singe vertices, essentially, between 2 nodes). Is this correct? If so, is there an opportunity to cluster segments to help create a higher level summary set of grouped route line components for spatial indexing?

Reply:
> Is this correct? 

Yep.

> If so, is there an opportunity to cluster segments to help create a higher level summary set of grouped route line components for spatial indexing?

Not sure I follow. How would this work?